custom/text-chat/contexts/ChatProvider.js

```diff
@@ -6,61 +6,73 @@ import React, {
   useState,
 } from 'react';
 import { useCallState } from '@custom/shared/contexts/CallProvider';
+import { useSharedState } from '@custom/shared/hooks/useSharedState';
 import { nanoid } from 'nanoid';
 import PropTypes from 'prop-types';
 
 export const ChatContext = createContext();
 
 export const ChatProvider = ({ children }) => {
   const { callObject } = useCallState();
-  const [chatHistory, setChatHistory] = useState([]);
+  const { sharedState, setSharedState } = useSharedState({
+    initialValues: {
+      chatHistory: [],
+    },
+    broadcast: false,
+  });
   const [hasNewMessages, setHasNewMessages] = useState(false);
 
   const handleNewMessage = useCallback(
     (e) => {
+      if (e?.data?.message?.type) return;
       const participants = callObject.participants();
       const sender = participants[e.fromId].user_name
         ? participants[e.fromId].user_name
         : 'Guest';
 
-      setChatHistory((oldState) => [
-        ...oldState,
-        { sender, message: e.data.message, id: nanoid() },
-      ]);
+      setSharedState({
+        ...sharedState,
+        chatHistory: [
+          ...sharedState.chatHistory,
+          { sender, senderID: e.fromId, message: e.data.message, id: nanoid() },
+        ]
+      });
```

Comment:
```suggestion
      setSharedState((oldState) => ({
        ... oldState,
        chatHistory: [
          ... oldState.chatHistory,
          { sender, senderID: e.fromId, message: e.data.message, id: nanoid() },
        ]
      }));
```
Suggestion: avoid the need to read from `sharedState`. Otherwise this would re-assign the `app-message` listener, whenever the `sharedState` changes (e.g. a chat message is received).

https://user-images.githubusercontent.com/1336590/141781985-cba42e2f-f2bd-4f29-94ec-ad0fdbac68f9.mp4